# Pandas Tutorial 
Pandas is a python library commonly used for analyzing, filtering and manipulating data. It works particularly well with any tabular ("table like") data, including the data sets that we will be using during Week 1. In this notebook, we will explore how pandas can be used to read, visualize, and modify data sets. For further information, see the pandas documentation website here: https://pandas.pydata.org/docs/index.html

## Pima Diabetes Dataset

This notebook tutorial and most of the other tutorial notebooks for this week will use data from the Pima Diabetes Dataset. The set we will use can be accessed with the following direct link:

https://raw.githubusercontent.com/BeaverWorksMedlytics2020/Data_Public/master/NotebookExampleData/Week1/diabetes.csv

### About the data:

All patients here are females at least 21 years old of Pima Native American heritage.

- **Pregnancies**: Number of times pregnant
- **Glucose**: Plasma glucose concentration 2 hours in an oral glucose tolerance test
- **BloodPressure**: Diastolic blood pressure (mm Hg)
- **SkinThickness**: Triceps skin fold thickness (mm)
- **Insulin**: 2-Hour serum insulin (mu U/ml)
- **BMI**: Body mass index (weight in kg/(height in m)^2)
- **DiabetesPedigreeFunction**: A function that scores the likelihood of diabetes based on family history.
- **Age**: Age (years)
- **Outcome**: 1 (has diabetes) or 0 (no diabetes)

## Reading CSV files with pandas

Pandas can read data directly from a url. There are a number of options associated with the **read_csv** function, but for now we will just pass a list of the column labels, since our CSV does not contain them. Without this option, the first row of data would be interpreted as labels.

In [1]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt

url = "https://raw.githubusercontent.com/BeaverWorksMedlytics2020/Data_Public/master/NotebookExampleData/Week1/diabetes.csv"

"""
      'preg': number of pregnancies  
      'plas': plasma glucose concentration 
      'pres': blood pressure 
      'skin': skin thickness
      'test': Insluin
      'mass': BMI
      'pedi': diabetes pedigree function
      'age': age
      'class': '0' means does not have diabetes and '1' means has diabetes
"""

# Define explicit list of column headers to pass to pandas
our_names = ['preg','plas','pres','skin','test','mass','pedi','age','class']
data = pd.read_csv(url, names=our_names)

Now that we've imported the data, let's take a closer look at the python object that was created by pandas. First, what is the object's type?

In [2]:
type(data)

pandas.core.frame.DataFrame

DataFrame is a very general structure for 2-dimensional tabular data. It contains column labels, is size-mutable, and may contain heterogeneous data (this means data may be missing, or of mixed types). There are a number of ways to get information about a DataFrame. Each of the following code segments will demonstrate one such way.

In [3]:
# Show the first 5 entries in the DataFrame (from the "head")
data.head(10)

preg  plas  pres  skin  test  mass   pedi  age  class
0     6   148    72    35     0  33.6  0.627   50      1
1     1    85    66    29     0  26.6  0.351   31      0
2     8   183    64     0     0  23.3  0.672   32      1
3     1    89    66    23    94  28.1  0.167   21      0
4     0   137    40    35   168  43.1  2.288   33      1
5     5   116    74     0     0  25.6  0.201   30      0
6     3    78    50    32    88  31.0  0.248   26      1
7    10   115     0     0     0  35.3  0.134   29      0
8     2   197    70    45   543  30.5  0.158   53      1
9     8   125    96     0     0   0.0  0.232   54      1

In [4]:
# Get the dimensionality of the DataFrame (number of rows, columns)
data.shape

(768, 9)

In [5]:
# Find the number of entries for each column
data.count()

preg     768
plas     768
pres     768
skin     768
test     768
mass     768
pedi     768
age      768
class    768
dtype: int64

In [6]:
# Create a string that describes the shape of the dataset
"There are " + str(data.shape[0]) + " people in the study with " + str(data.shape[1]) + " variables each "

'There are 768 people in the study with 9 variables each '

In [7]:
# Generate a table of statistical information describing the dataset
data.describe()

preg        plas        pres  ...        pedi         age       class
count  768.000000  768.000000  768.000000  ...  768.000000  768.000000  768.000000
mean     3.845052  120.894531   69.105469  ...    0.471876   33.240885    0.348958
std      3.369578   31.972618   19.355807  ...    0.331329   11.760232    0.476951
min      0.000000    0.000000    0.000000  ...    0.078000   21.000000    0.000000
25%      1.000000   99.000000   62.000000  ...    0.243750   24.000000    0.000000
50%      3.000000  117.000000   72.000000  ...    0.372500   29.000000    0.000000
75%      6.000000  140.250000   80.000000  ...    0.626250   41.000000    1.000000
max     17.000000  199.000000  122.000000  ...    2.420000   81.000000    1.000000

[8 rows x 9 columns]

Try using tab complete to see other functions and variables associated with DataFrames, or explore the API documentation:

https://pandas.pydata.org/docs/reference/api/pandas.DataFrame.html

## Slicing and Dicing

Slicing and dicing are technical terms for subselecting parts of a multi-dimensional dataset. To **slice** a dataset is to choose a value for one dimension, and take all data that matches. This results in a subset that is one dimension smaller. For example, in our dataset we could slice a single row or column. To **dice** data is to choose ranges in multiple dimensions, creating a more arbitrary subset.

To take a simple example, selecting a value from a python list by index is technically a slice (1-D to 0-D). Selecting a range of indices would be a dice. Pandas allows for similar data access, either by index or label. Here are some examples:
<br></br>

**iloc**: Indexing with integers. Accepts single values [ i ], lists [ i, j ], or ranges [ i : j : k ].

In [8]:
# Single-row slice
data.iloc[3]

preg      1.000
plas     89.000
pres     66.000
skin     23.000
test     94.000
mass     28.100
pedi      0.167
age      21.000
class     0.000
Name: 3, dtype: float64

In [9]:
# Dicing the first three rows
data.iloc[0:3]  # NOTE: equivalent to data.iloc[0:3:1]

preg  plas  pres  skin  test  mass   pedi  age  class
0     6   148    72    35     0  33.6  0.627   50      1
1     1    85    66    29     0  26.6  0.351   31      0
2     8   183    64     0     0  23.3  0.672   32      1

In [11]:
# 2-D dice of the first 3 rows, but only the last two columns
data.iloc[:3, 7:]  # NOTE: equivalent to data.iloc[0:3:1, 7:9:1]

age  class
0   50      1
1   31      0
2   32      1

**loc**: Indexing with column labels, which are strings for columns and integers for rows. Accepts single values [ 'pres' ], lists [ 'pres', 'class' ], or "label slices" [ 'pres' : 'class' ]. Note that the "label slice" is not a range, so the terminal value will be included. This is particularly confusing with row labels, because the syntax for a slice will resemble an integer range, as shown below.

In [12]:
# Dice the first three rows, with data in the columns from 'pres' to 'class'
data.loc[0:2, 'pres':'class'] # NOTE: 0:2 is a slice, not a range, and includes index 2

pres  skin  test  mass   pedi  age  class
0    72    35     0  33.6  0.627   50      1
1    66    29     0  26.6  0.351   31      0
2    64     0     0  23.3  0.672   32      1

## Chained Indexing/Masking

Now we know how to subselect data, but that's not very useful unless we know what data we want to subselect. One solution is to create a data **mask**. In this context, the mask is a list of indices corresponding to data that meets certain criteria. For example, we can create a mask for all data for people with an age greater than 39. Doing so is quite simple in pandas:

In [13]:
# Create a pandas series "mask" for rows with an age greater than 39
data_age_boolean = data.age > 39
# Show the created boolean series
data_age_boolean.head(20)

0      True
1     False
2     False
3     False
4     False
5     False
6     False
7     False
8      True
9      True
10    False
11    False
12     True
13     True
14     True
15    False
16    False
17    False
18    False
19    False
Name: age, dtype: bool

The created object is a series of True/False values for each row. Let's double check the first three values for age.

In [14]:
# Show the first three age values. Only value 0 is greater than 39.
data['age'].head(3)

0    50
1    31
2    32
Name: age, dtype: int64

In [15]:
# Show the first three values in the boolean mask. Only value 0 should be TRUE.
data_age_boolean.head(3)

0     True
1    False
2    False
Name: age, dtype: bool

The series object can be used to access a subset of the data. This is effectively filtering the entire dataset to return only entries with an age greater than 39.

In [16]:
# Subselect data that meets the condition set by data_age_boolean
filtered_dataframe_by_age = data[data_age_boolean]
# Display the resulting filtered data
filtered_dataframe_by_age

preg  plas  pres  skin  test  mass   pedi  age  class
0       6   148    72    35     0  33.6  0.627   50      1
8       2   197    70    45   543  30.5  0.158   53      1
9       8   125    96     0     0   0.0  0.232   54      1
12     10   139    80     0     0  27.1  1.441   57      0
13      1   189    60    23   846  30.1  0.398   59      1
..    ...   ...   ...   ...   ...   ...    ...  ...    ...
757     0   123    72     0     0  36.3  0.258   52      1
759     6   190    92     0     0  35.5  0.278   66      1
761     9   170    74    31     0  44.0  0.403   43      1
763    10   101    76    48   180  32.9  0.171   63      0
766     1   126    60     0     0  30.1  0.349   47      1

[207 rows x 9 columns]

Notice that while the filtered output is only 207 rows, the row label goes up to 766. This is because the subselected rows have kept their original labels. Row "766" is the 207th row in our new structure. This can be important if we want to refer back to the original dataset.

Often we will want to apply more than one filter to our data. Fortunately, since the masks are just booleans, it is intuitive and easy to combine masks. For example, if we want to combine our age filter with a second filter, we can use a logical "and" operator (&) like this:

In [17]:
# Subselect data with age greater than 39 and more than 2 pregnancies
filtered_dataframe_age_preg = data[data_age_boolean & (data['preg'] > 2)]
# Display the filtered data
filtered_dataframe_age_preg

preg  plas  pres  skin  test  mass   pedi  age  class
0       6   148    72    35     0  33.6  0.627   50      1
9       8   125    96     0     0   0.0  0.232   54      1
12     10   139    80     0     0  27.1  1.441   57      0
14      5   166    72    19   175  25.8  0.587   51      1
21      8    99    84     0     0  35.4  0.388   50      0
..    ...   ...   ...   ...   ...   ...    ...  ...    ...
749     6   162    62     0     0  24.3  0.178   50      1
754     8   154    78    32     0  32.4  0.443   45      1
759     6   190    92     0     0  35.5  0.278   66      1
761     9   170    74    31     0  44.0  0.403   43      1
763    10   101    76    48   180  32.9  0.171   63      0

[172 rows x 9 columns]

Now we've reduced our subset to just 172 entries. Also, while the original row labels are still maintained, row "766" has been filtered out!

As a final note, you can also combine filters with the "or" operator (|). We would expect this to produce a subset of equal or larger size than either filter on its own.

In [18]:
# Subselect data with age greater than 39 OR more than 2 pregnancies
filtered_dataframe_age_preg = data[data_age_boolean | (data['preg'] > 2)]
# Display the filtered data
filtered_dataframe_age_preg

preg  plas  pres  skin  test  mass   pedi  age  class
0       6   148    72    35     0  33.6  0.627   50      1
2       8   183    64     0     0  23.3  0.672   32      1
5       5   116    74     0     0  25.6  0.201   30      0
6       3    78    50    32    88  31.0  0.248   26      1
7      10   115     0     0     0  35.3  0.134   29      0
..    ...   ...   ...   ...   ...   ...    ...  ...    ...
761     9   170    74    31     0  44.0  0.403   43      1
762     9    89    62     0     0  22.5  0.142   33      0
763    10   101    76    48   180  32.9  0.171   63      0
765     5   121    72    23   112  26.2  0.245   30      0
766     1   126    60     0     0  30.1  0.349   47      1

[454 rows x 9 columns]

You can also sort the values in a dataframe by ascending or descending order. This can be useful for quickly viewing, for example, all of the subjects who have had more than 12 pregnancies.

In [25]:
sample = filtered_dataframe_age_preg.sort_values(by='preg', ascending=False, na_position='first') #descending order
sample #ordered based on "preg"

preg  plas  pres  skin  test  mass   pedi  age  class
159    17   163    72    41   114  40.9  0.817   47      1
88     15   136    70    32   110  37.1  0.153   43      1
455    14   175    62    30     0  33.6  0.212   38      1
298    14   100    78    25   184  36.6  0.412   46      1
635    13   104    72     0     0  31.2  0.465   38      1
..    ...   ...   ...   ...   ...   ...    ...  ...    ...
757     0   123    72     0     0  36.3  0.258   52      1
294     0   161    50     0     0  21.9  0.254   65      0
129     0   105    84     0     0  27.9  0.741   62      1
58      0   146    82     0     0  40.5  1.781   44      0
440     0   189   104    25     0  34.3  0.435   41      1

[454 rows x 9 columns]

# Dealing with missing data

Not all datasets are created whole. Survey respondents may accidentally skip a question, some lab equipement could malfunction, or the handwriting for an answer may be illegible to the people entering the data into a computer. Anyway it could happen, and missing data is annoying!

There are a few ways to go about fixing those fields in your data, each with their own advantages and disadvantages. We'll continue using the Pima diabetes dataset to demonstrate.

In [28]:
data.head(15)

preg   plas  pres  skin   test  mass   pedi  age  class
0      6  148.0  72.0  35.0    NaN  33.6  0.627   50      1
1      1   85.0  66.0  29.0    NaN  26.6  0.351   31      0
2      8  183.0  64.0   NaN    NaN  23.3  0.672   32      1
3      1   89.0  66.0  23.0   94.0  28.1  0.167   21      0
4      0  137.0  40.0  35.0  168.0  43.1  2.288   33      1
5      5  116.0  74.0   NaN    NaN  25.6  0.201   30      0
6      3   78.0  50.0  32.0   88.0  31.0  0.248   26      1
7     10  115.0   NaN   NaN    NaN  35.3  0.134   29      0
8      2  197.0  70.0  45.0  543.0  30.5  0.158   53      1
9      8  125.0  96.0   NaN    NaN   NaN  0.232   54      1
10     4  110.0  92.0   NaN    NaN  37.6  0.191   30      0
11    10  168.0  74.0   NaN    NaN  38.0  0.537   34      1
12    10  139.0  80.0   NaN    NaN  27.1  1.441   57      0
13     1  189.0  60.0  23.0  846.0  30.1  0.398   59      1
14     5  166.0  72.0  19.0  175.0  25.8  0.587   51      1

Now if you look into the dataframe you'll see some odd, or rather **missing**, values. We can tell that these are missing values from knowing what the columns stand for; for example it wouldn't make sense for someone to have a blood pressure or BMI of 0 right? 

In [27]:
# Looking at the columns with 0s
data.describe()

preg        plas        pres  ...        pedi         age       class
count  768.000000  763.000000  733.000000  ...  768.000000  768.000000  768.000000
mean     3.845052  121.686763   72.405184  ...    0.471876   33.240885    0.348958
std      3.369578   30.535641   12.382158  ...    0.331329   11.760232    0.476951
min      0.000000   44.000000   24.000000  ...    0.078000   21.000000    0.000000
25%      1.000000   99.000000   64.000000  ...    0.243750   24.000000    0.000000
50%      3.000000  117.000000   72.000000  ...    0.372500   29.000000    0.000000
75%      6.000000  141.000000   80.000000  ...    0.626250   41.000000    1.000000
max     17.000000  199.000000  122.000000  ...    2.420000   81.000000    1.000000

[8 rows x 9 columns]

You can see from the 4th row showing minimum values that 'preg', 'plas', 'pres', 'skin', 'test', 'mass', and 'class' have 0s. On the other hand don't forget that for columns like number of pregnancies and class, a 0 is perfectly valid.

_This is why it's important that you know what type of data you're handling!_

To summarize, the columns with invalid 0s (missing data) are:
- plas
- pres
- skin
- test
- mass

## Replacing with a constant

A lot of times when data is missing the cell is simply left blank or filled with NaNs or constant; in our case **_they have already been replaced with the constant 0_**. However as said before we want to keep the 0s in columns representing pregnancies and class so it'd be best to somehow differentiate these missing and valid values.

NaN stands for "Not a Number" and is usually used for missing entries in a dataframe. Let's go ahead and replace the invalid 0s with NaNs.

In [26]:
invalid = ['plas', 'pres', 'skin', 'test', 'mass']

for i in invalid:
    data[i].replace(to_replace=0, value=np.nan, inplace=True)

data.head(15)

preg   plas  pres  skin   test  mass   pedi  age  class
0      6  148.0  72.0  35.0    NaN  33.6  0.627   50      1
1      1   85.0  66.0  29.0    NaN  26.6  0.351   31      0
2      8  183.0  64.0   NaN    NaN  23.3  0.672   32      1
3      1   89.0  66.0  23.0   94.0  28.1  0.167   21      0
4      0  137.0  40.0  35.0  168.0  43.1  2.288   33      1
5      5  116.0  74.0   NaN    NaN  25.6  0.201   30      0
6      3   78.0  50.0  32.0   88.0  31.0  0.248   26      1
7     10  115.0   NaN   NaN    NaN  35.3  0.134   29      0
8      2  197.0  70.0  45.0  543.0  30.5  0.158   53      1
9      8  125.0  96.0   NaN    NaN   NaN  0.232   54      1
10     4  110.0  92.0   NaN    NaN  37.6  0.191   30      0
11    10  168.0  74.0   NaN    NaN  38.0  0.537   34      1
12    10  139.0  80.0   NaN    NaN  27.1  1.441   57      0
13     1  189.0  60.0  23.0  846.0  30.1  0.398   59      1
14     5  166.0  72.0  19.0  175.0  25.8  0.587   51      1

**Pros**
- easiest and quickest way to deal with missing values


**Cons**
- can introduce bias and skew data

## Dropping rows

One of the next easiest ways to deal with NaNs is simply deleting rows with missing information. Luckily, pandas already has a beautiful function just for that purpose.

In [29]:
data_row = data.dropna(axis=0).reset_index(drop=True) # axis: 0=row, 1=column

data_row.head(15)

preg   plas  pres  skin   test  mass   pedi  age  class
0      1   89.0  66.0  23.0   94.0  28.1  0.167   21      0
1      0  137.0  40.0  35.0  168.0  43.1  2.288   33      1
2      3   78.0  50.0  32.0   88.0  31.0  0.248   26      1
3      2  197.0  70.0  45.0  543.0  30.5  0.158   53      1
4      1  189.0  60.0  23.0  846.0  30.1  0.398   59      1
5      5  166.0  72.0  19.0  175.0  25.8  0.587   51      1
6      0  118.0  84.0  47.0  230.0  45.8  0.551   31      1
7      1  103.0  30.0  38.0   83.0  43.3  0.183   33      0
8      1  115.0  70.0  30.0   96.0  34.6  0.529   32      1
9      3  126.0  88.0  41.0  235.0  39.3  0.704   27      0
10    11  143.0  94.0  33.0  146.0  36.6  0.254   51      1
11    10  125.0  70.0  26.0  115.0  31.1  0.205   41      1
12     1   97.0  66.0  15.0  140.0  23.2  0.487   22      0
13    13  145.0  82.0  19.0  110.0  22.2  0.245   57      0
14     3  158.0  76.0  36.0  245.0  31.6  0.851   28      1

But be cautious with dropping rows, depending on how many rows had NaNs and the size of your data. You may inadvertently drastically decrease the sample size. As you can see below, we lost about half of our data:

In [30]:
print("Original size: ", data.shape)
print("With rows dropped: ", data_row.shape)

Original size:  (768, 9)
With rows dropped:  (392, 9)


**Pros**
- again, easy and quick to apply to data

**Cons**
- reduces sample size

## Dropping the variable (column)

If the majority of the column were filled with NaNs (ie >70% but it's up to you to determine your own rule of thumb) and thus not very important in predicting the targeted label, dropping the variable makes sense. Although overall, it would be unwise to drop a variable just because it has NaNs; as you can see in our data that would mean dropping 5 of our 9 columns!

For an example's sake, let's drop the column containing the most NaNs.

In [31]:
for i in invalid:
    count = data[i].isna().sum()
    print(i, ": ", count, "NaNs, ", ((count/(len(data.index))) * 100), "%")

plas :  5 NaNs,  0.6510416666666667 %
pres :  35 NaNs,  4.557291666666666 %
skin :  227 NaNs,  29.557291666666668 %
test :  374 NaNs,  48.69791666666667 %
mass :  11 NaNs,  1.4322916666666665 %


In [32]:
data_col = data.drop(['test'], axis=1)

data_col.head()

preg   plas  pres  skin  mass   pedi  age  class
0     6  148.0  72.0  35.0  33.6  0.627   50      1
1     1   85.0  66.0  29.0  26.6  0.351   31      0
2     8  183.0  64.0   NaN  23.3  0.672   32      1
3     1   89.0  66.0  23.0  28.1  0.167   21      0
4     0  137.0  40.0  35.0  43.1  2.288   33      1

**Pros**
- can clean up unnecessary data and speed up processing

**Cons**
- reduces data and number of potential features

## Replacing with the mean/median

We can also deal with missing values by replacing them with the mean or median if the column's numerical. Like dropping NaNs, filling NaNs with a constant value can be done with a single function. 

In [33]:
for i in invalid:
    data[i].fillna(data[i].mean(), inplace=True) #use data[i].median() for median

data.head()

preg   plas  pres      skin        test  mass   pedi  age  class
0     6  148.0  72.0  35.00000  155.548223  33.6  0.627   50      1
1     1   85.0  66.0  29.00000  155.548223  26.6  0.351   31      0
2     8  183.0  64.0  29.15342  155.548223  23.3  0.672   32      1
3     1   89.0  66.0  23.00000   94.000000  28.1  0.167   21      0
4     0  137.0  40.0  35.00000  168.000000  43.1  2.288   33      1

**Pros**
- easy and straightforward (calculate, find, & replace)

**Cons**
- skews data to be biased towards the center, underestimating the variance in our data 